Feat/2486 - Parser Performance Improvement

elipe17 opened the pull request:
## Summary of Changes
- Update parsing logic to avoid multiple passes over datafile
- Update record saving to occur in batches of 10,000 instead of one at a time
Pull request closes #2486 

## How to Test

```
cd tdrs-frontend && docker-compose up
cd tdrs-backend && docker-compose up
cd tdrs-backend && docker-compose exec web pytest -k test_parse.py
```
- You could also verify the time decrease by running
```
cd tdrs-backend && docker-compose exec web pytest -k test_parse_ssp_section1_datafile
```
on both `develop` and this branch. On my machine, develop takes ~820 seconds to complete and this branch takes ~16 seconds to complete. Please use the attached patch to add timing code to this test. Note the test is forced to fail so that you can see the stdout of the time taken in seconds. You can apply the patch by:
```
cd tdrs-backend && git apply timer.patch
```

[timer.patch](https://github.com/raft-tech/TANF-app/files/11580133/timer.patch)


## Deliverables
_More details on how deliverables herein are assessed included [here](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverables)._

### [Deliverable 1: Accepted Features](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-1-Accepted-Features)

Checklist of ACs:
- [x] Identify root cause of any performance issues
- [x] Identify mitigation steps
- [x] Create a benchmarking test to compare and contrast current performance with any possible improvements
- [x] Testing Checklist has been run and all tests pass
- [x] README is updated, if necessary

### [Deliverable 2: Tested Code](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-2-Tested-Code)

+ Are all areas of code introduced in this PR meaningfully tested?
  + [x] If this PR introduces backend code changes, are they meaningfully tested?
  + [x] If this PR introduces frontend code changes, are they meaningfully tested?
+ Are code coverage minimums met?
  + [x] Frontend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)
  + [x] Backend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)

### [Deliverable 3: Properly Styled Code](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-3-Properly-Styled-Code)

+ [x] Are backend code style checks passing on CircleCI?
+ [x] Are frontend code style checks passing on CircleCI?
+ [x] Are code maintainability principles being followed?

### [Deliverable 4: Accessible](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-4-Accessibility)

+ [ ] Does this PR complete the epic? 
+ [ ] Are links included to any other gov-approved PRs associated with epic?
+ [ ] Does PR include documentation for Raft's a11y review? 
+ [ ] Did automated and manual testing with `iamjolly` and `ttran-hub` using Accessibility Insights reveal any errors introduced in this PR?


### [Deliverable 5: Deployed](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-5-Deployed)

+ [x] Was the code successfully deployed via automated CircleCI process to development on Cloud.gov?

### [Deliverable 6: Documented](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-6-Code-documentation)

+ [x] Does this PR provide background for why coding decisions were made?
+ [x] If this PR introduces backend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces frontend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces dependencies, are their licenses documented?
+ [x] Can reviewer explain and take ownership of these elements presented in this code review?

### [Deliverable 7: Secure](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-7-Secure)

+ [x] Does the OWASP Scan pass on CircleCI?
+ [ ] Do manual code review and manual testing detect any new security issues?
+ [ ] If new issues detected, is investigation and/or remediation plan documented? 

### [Deliverable 8: User Research](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-8-User-Research)

Research product(s) clearly articulate(s):
+ [ ] the purpose of the research
+ [ ] methods used to conduct the research 
+ [ ] who participated in the research
+ [ ] what was tested and how
+ [ ] impact of research on TDP
+ [ ] (_if applicable_) final design mockups produced for TDP development

elipe17 committed:
- ADding t2 schema

elipe17 committed:
- small fixes to naming
- added t2.py
- Updating var names to match field names

elipe17 committed:
Merge branch 'develop' of https://github.com/raft-tech/TANF-app into 2424-ts1-parsing

elipe17 committed:
- Added new doc for T2
- Updated model fo rT2
- Added active parse function

elipe17 committed:
- Added T3 schema defs

elipe17 committed:
- Fixing most lint errors

elipe17 committed:
- Updated T3 to multirow record

elipe17 committed:
- Added unit test and data files
- updated field names

elipe17 committed:
- Updating var names to conform to DB max var name length

elipe17 committed:
- Updating based on unit test failure

elipe17 committed:
- adding datafile

elipe17 committed:
- Adding unit tests for t2 and t3

elipe17 committed:
Merge branch 'develop' of https://github.com/raft-tech/TANF-app into feature/1113-1-ssp-section-1-models

elipe17 committed:
- Fixed field types
- Updated test file

elipe17 committed:
- Removing many migrations to consolodate into one migration

elipe17 committed:
Merge branch 'develop' of https://github.com/raft-tech/TANF-app into 2424-ts1-parsing

elipe17 committed:
Revert "- Removing many migrations to consolodate into one migration"

This reverts commit 1455ae4d334f72e250405803d61d26c0a894e886.

elipe17 committed:
- Resolved test issues after merge

elipe17 committed:
- Resolve lint errors

elipe17 committed:
Merge branch 'feature/1113-1-ssp-section-1-models' of https://github.com/raft-tech/TANF-app into feature/1113-1-ssp-section-1-models

elipe17 committed:
Merge branch 'feature/1113-1-ssp-section-1-models' into 2424-ts1-parsing

elipe17 committed:
- Merged 1113.2

elipe17 committed:
Merge branch 'feature/1113-2-ssp-section-1-parsing' of https://github.com/raft-tech/TANF-app into 2424-ts1-parsing

elipe17 committed:
- Removed unnecessary file

elipe17 committed:
- Rename model fields
- Combined migrations

elipe17 committed:
- fixed spelling in test

elipe17 committed:
- Fixed lint errors

jtimpe committed:
rm commented validators

elipe17 committed:
- Updated schema newlines to be consistent

ADPennington committed:
Merge branch 'develop' into feature/1113-2-ssp-section-1-parsing

elipe17 committed:
Merge branch 'feature/1113-2-ssp-section-1-parsing' of https://github.com/raft-tech/TANF-app into 2424-ts1-parsing

elipe17 committed:
- Updated field names
- Updated tests
- Updated migration

elipe17 committed:
Merge branch 'develop' of https://github.com/raft-tech/TANF-app into 2424-ts1-parsing

elipe17 committed:
- consolodating migrations

elipe17 committed:
- Updated readme and backup script

elipe17 committed:
Merge branch 'develop' of https://github.com/raft-tech/TANF-app into 2424-ts1-parsing

elipe17 committed:
- Updated parse logic to batch create records

elipe17 committed:
- Fixed lint errors

elipe17 committed:
- Batching record serialization

elipe17 committed:
- Updated parse logic

elipe17 committed:
- Adding big s1 datafile

elipe17 committed:
- fixing lint errors

elipe17 committed:
Merge branch 'develop' of https://github.com/raft-tech/TANF-app into 2486-parser-performance

raftmsohani commented on tdrs-backend/tdpservice/parsers/parse.py:
```diff
@@ -61,27 +41,69 @@ def parse_datafile(datafile):
+    """Store record in dictionary for later processing."""
+    if record:
+        if model not in unsaved_records:
+            unsaved_records[model] = [record]
```
Nit: to make it more readable, always prefer to do something like below if unsaved_records is doctionary:

unsaved_records.update({
    model: record
    })

raftmsohani commented on tdrs-backend/tdpservice/parsers/parse.py:
```diff
@@ -61,27 +41,69 @@ def parse_datafile(datafile):
+        if model not in unsaved_records:
+            unsaved_records[model] = [record]
+        else:
+            unsaved_records[model].append(record)
```
But I see here that unsaved_records is not dictionary?? 
Looks like the method docstring is not correct since it reads the result is dictionary?

elipe17 replied:
unsaved records is a dict, but its keys point to a list of records to be serialized. I can split that up to make it less confusing.

raftmsohani commented on tdrs-backend/tdpservice/parsers/parse.py:
```diff
@@ -61,27 +41,69 @@ def parse_datafile(datafile):
+        errors['trailer'] = ['Multiple trailers found.']
+        multiple_trailer_errors = True
+    if trailer_count == 1:
+        _, trailer_is_valid, trailer_errors = schema_defs.trailer.parse_and_validate(line)
```
Please assign var "header" to make it clear what '_' is referring to

raftmsohani commented on tdrs-backend/tdpservice/parsers/parse.py:
```diff
@@ -61,27 +41,69 @@ def parse_datafile(datafile):
+        evaluate_trailer(trailer_count, multiple_trailer_errors, line, errors)
+
+        if header_count > 1:
+            errors['document'] = ['Multiple headers found.']
```
Nit: would use .update()
``` 
errors.update({
    'document': 'Mul....'
    })
```

elipe17 committed:
- Removing test to see if its causing weird failure

elipe17 committed:
- Updating based on comments
- Removing big file since circi cant handle it

elipe17 committed:
- fixing error in update method

elipe17 committed:
- fixing error in update

elipe17 committed:
- Removed extraneous seek

elipe17 committed:
- updating ignore to ignore patch files

jtimpe commented on tdrs-backend/tdpservice/parsers/parse.py:
```diff
@@ -61,27 +41,71 @@ def parse_datafile(datafile):
+    """Store record in dictionary for later processing."""
+    if record:
+        if model not in unsaved_records:
+            unsaved_records.update({model: [record]})
```
`store_record`, as well as `evaluate_trailer` below, break both the purity and immutability principles of functional programming because they modify one of the passed-in parameters (`unsaved_records` in this case). A slightly better solution might be to build a new `unsaved_records` object based on a copy of the original and return it, overriding the value in the calling function.

```python
def store_record(unsaved_records, record, model):
    new_records = unsaved_records.copy()
    if record:
        new_records.get(model, []).append(record)  # might work as a 1-line replacement of lines 55-58
    return new_records


## then, inside `parse_datafile_lines`

unsaved_records = store_record(unsaved_records, record, model)  # override the previously set value
```

elipe17 replied:
For the store record that becomes really expensive. Dictionary.copy() is O(n) (n is the number of records) time complexity and we would make the copy n times. That would give the overall complexity of the parse_datafile_lines an O(n^2) complexity which hurts a lot. I can do some testing to see how much it hurts to verify.

jtimpe commented on tdrs-backend/tdpservice/parsers/parse.py:
```diff
@@ -61,27 +41,71 @@ def parse_datafile(datafile):
+
+        if header_count > 1:
+            errors.update({'document': ['Multiple headers found.']})
+            return errors
```
so, i see what you were going for here - this avoids having to run through the file twice in order to do the "single header/trailer" pre-check, however i think this may result in the unintended effect of allowing every record up until the duplicated header/trailer to be parsed as well, whereas during our initial conversations i recall us wanting to handle it in a different step (so that we would not attempt *any* parsing on a document with multiple headers/trailers). not opposed to this change, just wanted to call out how it changes the parser behavior.

i am curious though, after implementing the `bulk_create` - how much time is actually saved by making the parsing process one loop instead of two? i'd imagine it's negligible compared to the database efficiency change.

elipe17 replied:
Ya I definitely wanted feedback on this. On all our existing datafiles it made little difference. However, on the 50MB file it saved ~28 seconds on my machine. That is why I left it in since it was more significant. But the main time killer was definitely the IO so removing this would not be a huge time loss if it is deemed better to die earlier wrt multiple headers/trailers.

jtimpe reviewed (changes requested):
This is really good! Parsing is lightning fast even for our biggest files. I'd like you to consider making a couple of the helper functions pure/immutable and I had an open question/observation about the header/trailer "precheck".

Lastly, when 1610.1 merges, there will be a new set of `ParserError` objects also created within the parser. I don't believe your `bulk_create` changes will handle this automatically, but hopefully it's not too much additional effort - the concept generally will be the same. However, since `ParserError` objects can have a relation to any of the document objects created in the parser, I'm curious how that will work with `bulk_create`. Will it be able to create the foreign key when the document object hasn't been officially created yet? I'd imagine we could work around that by setting a `uuid` pk application-side while creating document objects, then the `ParserError` would know the record's `id`. Food for thought, though.

andrew-jameson commented on tdrs-backend/tdpservice/parsers/parse.py:
```diff
@@ -61,27 +41,71 @@ def parse_datafile(datafile):
+    if trailer_count == 1:
+        record, trailer_is_valid, trailer_errors = schema_defs.trailer.parse_and_validate(line)
+        if not trailer_is_valid:
+            errors['trailer'] = trailer_errors
```
Surely we'd want to return errors and catch it in `parse_datafile_lines()`?

andrew-jameson commented on tdrs-backend/tdpservice/parsers/parse.py:
```diff
@@ -90,17 +114,29 @@ def parse_datafile_lines(datafile, program_type, section):
+                errors.update({line_number: record_errors})
+            store_record(unsaved_records, record, schema.model)
+
+        if line_number % 50000 == 0 and header_count > 0:
```
Not a fan of this 50k magic number floating here. Could the batch size be regulated inside one of our functions either the "store records" or "bulk create"? Maybe once the dict in store records hits our variable `batch_size`, we save it? That might be slightly more readable.

elipe17 committed:
- Made evaluate_trailer pure/immutable

elipe17 committed:
Revert "- Removing test to see if its causing weird failure"

This reverts commit 64b78b737c97fb9bfa70217ff70ccffea8d85429.

elipe17 committed:
- Use custom transactions while parsing
- Handle transaction rollback on parse failure/error
- Update tests to use transaction logic
- Created new decorator to rollback database when assertion error is hit
- Making elastic search log level configurable
- Added test for inter parse rollback

elipe17 committed:
- updated store logic to one liner
- updated decorator to catch all exceptions
- updated tests

elipe17 committed:
- removed function
- renamed test

elipe17 committed:
Merge branch 'develop' of https://github.com/raft-tech/TANF-app into 2486-parser-performance

elipe17 committed:
- refactored multirecord schema to be a schema manager instead
- updated parse functions to use the new layout
- updated bulk create to manage batch size

elipe17 committed:
- Name update for clarity

elipe17 committed:
Merge branch 'develop' of https://github.com/raft-tech/TANF-app into 2486-parser-performance
- Fixed conflicts
- Updated tests to reflect new order of errors

elipe17 committed:
- Fix lint errors

elipe17 committed:
- Changing batch size to avoid crash in circi

jtimpe commented on tdrs-backend/tdpservice/parsers/parse.py:
```diff
@@ -1,49 +1,29 @@
     rawfile = datafile.file
     errors = {}
 
-    document_is_valid, document_error = validators.validate_single_header_trailer(datafile)
```
can we delete `validators.validate_single_header_trailer` since it's no longer used?

jtimpe commented on tdrs-backend/tdpservice/parsers/parse.py:
```diff
@@ -65,97 +47,152 @@ def parse_datafile(datafile):
+    if trailer_count == 1 or is_last_line:
+        record, trailer_is_valid, trailer_errors = schema_defs.trailer.parse_and_validate(
+            line,
+            util.make_generate_parser_error(datafile, -1)
```
do you have access to the actual line number at this point? can we include that in the error instead of `-1`?

jtimpe commented on tdrs-backend/tdpservice/parsers/parse.py:
```diff
@@ -65,97 +47,152 @@ def parse_datafile(datafile):
+
+        if header_count > 1:
+            errors.update({'document': ['Multiple headers found.']})
+            err_obj = util.make_generate_parser_error(datafile, line_number)(
```
`util.make_generate_parser_error` is called a few times in the loop. since it's a higher order function, you can call it once at the start of the loop

```python
generate_error = util.make_generate_parser_error(datafile, line_number)
```

then use/pass `generate_error` throughout the rest of the loop. in one-off cases, you also might just want to use `util.generate_parser_error` which will let you pass in the datafile and line number along with the rest of the arguments

jtimpe commented on tdrs-backend/tdpservice/parsers/util.py:
```diff
@@ -255,7 +259,32 @@ def parse_and_validate(self, line, generate_error):
+
+def begin_transaction():
+    """Create manual database transaction."""
+    transaction.set_autocommit(False)
```
these feel a little funny to me because they just use an imported `db.transaction`. i would expect some sort of transaction per-request, that you'd have to pass into these util functions. instead, it seems like there's one long-running transaction where you turn `autocommit` on/off. am i understanding that correctly? and are there any implications to this that you're aware of?

jtimpe reviewed (approved):
This works really well! Couple small nitpick-y comments and one discussion point but i don't feel like it needs to block the merge. Great job!

elipe17 committed:
- Updated based on review
- Updated tests to reflect line number

raftmsohani reviewed (approved):
LGTM!

elipe17 committed:
- Updated based on review/OH
- Added extra transaction safety

elipe17 commented:
Let me know what you think about how I am handling the potential of a database error on the bulk create. I couldnt think of a nicer way without introducing "complexity" lint errors.

elipe17 committed:
Merge branch 'develop' of https://github.com/raft-tech/TANF-app into 2486-parser-performance

jtimpe commented on tdrs-backend/tdpservice/parsers/parse.py:
```diff
@@ -1,49 +1,33 @@
     )
     if not header_is_valid:
         errors['header'] = header_errors
+        ParserError.objects.bulk_create(header_errors)
```
one more place you might want to use `bulk_create_errors`

elipe17 committed:
- Updated view to show latest datafiles
- Added admin filter to show newest or all datafile records
- Updated indices to allow easier elastic queries

elipe17 committed:
- Updated based on review  comments

jtimpe reviewed (approved):
quick note and maybe @ADPennington and @andrew-jameson can weigh in with opinions here - uploading the `fake_rollback` test file included in this pr results in the following error spreadsheet

![image](https://github.com/raft-tech/TANF-app/assets/111305129/3d556cc2-fdf0-40b6-a5c1-1dfec9808e31)

this file contains a duplicated header at line 50,022 - after the 50k batch limit - but also contains a record field error before the batch limit. because of the multiple header error, all records created in the first 50k batch are rolled back, but the parser errors created for any of those records still exist. Eric and I chatted a bit about this and we both could go either way on this: is it better to
* also roll back any `ParserErrors` created for records that are rolled back
* leave the `ParserError` objects for rolled-back records so that Data Analysts can still see any error information for the lines in the file

I think this ambiguity come about because we've blurred the definition of "pre-parsing" with this change. Since we're only doing one file pass, we *can* accumulate record errors for records up until a "pre-parsing" error cancels the file parsing and rolls back any created records.

@elipe17 please correct me if i got anything wrong

ADPennington commented:
> o

interesting question @jtimpe!  for all practical purposes, I would assume that STTs would only make the corrections to rows that had the errors and not try to check for similar issues in rows that werent checked by the system (although that would be good practice). In other words, I'm somewhat dubious that seeing some of the errors (up to where the header issue is detected) will help them fix all possible errors on the re-submission. It could be more efficient if they got all the errors of a particular type at one time. 

I'm still leaning toward stop checking the file if a pre-parsing issue is detected, but i am open to suggestions that would keep the parsing logic simple. Also, not too long ago, @reitermb mentioned the value of checking the record types before parsing, and I believe that could help us detect the multiple headers issue sooner?

ADPennington commented:
> > o
> 
> interesting question @jtimpe! for all practical purposes, I would assume that STTs would only make the corrections to rows that had the errors and not try to check for similar issues in rows that werent checked by the system (although that would be good practice). In other words, I'm somewhat dubious that seeing some of the errors (up to where the header issue is detected) will help them fix all possible errors on the re-submission. It could be more efficient if they got all the errors of a particular type at one time.
> 
> I'm still leaning toward stop checking the file if a pre-parsing issue is detected, but i am open to suggestions that would keep the parsing logic simple. Also, not too long ago, @reitermb mentioned the value of checking the record types before parsing, and I believe that could help us detect the multiple headers issue sooner?

@jtimpe @elipe17 @andrew-jameson @reitermb let's take 10 min after a standup this week to discuss this. we can also take time during sprint ceremonies (if it ends early) @Smithh-Co

elipe17 committed:
- Updated based on feedback for preparse error handling
- updated tests

elipe17 committed:
- Updated search indices to have parent FK

elipe17 committed:
- Fix lint errors

elipe17 committed:
- Updated submission tests
- Moved create_datafile to util

elipe17 committed:
- fix lint errors

elipe17 committed:
- removing frontend filtering

raftmsohani commented on tdrs-backend/tdpservice/parsers/parse.py:
```diff
@@ -65,97 +51,170 @@ def parse_datafile(datafile):
+    """Bulk create passed in records."""
+    if (line_number % batch_size == 0 and header_count > 0) or flush:
+        try:
+            num_created = 0
```
nit: num_records_expected/created

elipe17 committed:
- addding datafile to admin model

elipe17 committed:
Revert "- addding datafile to admin model"

This reverts commit 35a6f24c36c3a4c00ddcfc40f20833530b0199f4.

elipe17 committed:
- Fixed issue where datafile FK wasnt populating
- Regenerated migration

elipe17 committed:
- Readding datafile back to admin view now that the error is resolved

elipe17 committed:
- adding datafile back

raftmsohani commented on tdrs-backend/tdpservice/parsers/parse.py:
```diff
@@ -65,97 +51,170 @@ def parse_datafile(datafile):
 
+    # Note: it is unnecessary to call rawfile.seek(0) again because the generator
+    # automatically starts back at the begining of the file.
+    begin_transaction(transaction)
```
I have two problems with "begin_transaction":
- I am very much leaned to wrapping it in try/except and turn transaction back in exception
- I am not clear what is the impact on other web tranactions while the autocommit is turned off. For example,  what if another user submits a file while autocommit is off? Is autocommit turned off for that transaction too?

elipe17 replied:
- We could certainly wrap the entirety of `parse_datafile_lines` in a try/except to avoid some random exception that might crop up. 
- As for the impact of transactions. From what I understand in the docs, transactions are scoped per process and per DB connection. So if another datafile is submitted and another worker/process executes it then there is no nesting/collision of transactions. If the same process is executing multiple calls to `parse_datafile_lines` and there arent unique DB connections, we will end up with nested transactions. Nested transactions are well supported however, if the first transaction is rolled back (i.e. the first datafile had some error where we deem we need to rollback) then any nested transaction would be rolled back to (this would include the second datafile).

raftmsohani replied:
Is there any docs online to reference: "transactions are scoped per process and per DB connection."? Would like to read more about this

elipe17 replied:
There arent specific docs surrounding that statement. It is just what I gathered from [this](https://docs.djangoproject.com/en/4.2/topics/db/transactions/ ) and [this](https://www.postgresql.org/docs/current/tutorial-transactions.html).

raftmsohani reviewed (commented):
see my comments

elipe17 committed:
Revert "- Readding datafile back to admin view now that the error is resolved"

This reverts commit 2807425059fd1b5b355edfb16d30d170cf869d7b.

elipe17 committed:
- Removed unnecessary fields
- Updated dependencies
- Updated filter

elipe17 committed:
- Updated document to include required fields

elipe17 committed:
- Moved datafile reference to avoid confusion

ADPennington committed:
Merge branch 'develop' into 2486-parser-performance

ADPennington commented:
@elipe17 testing notes below. I tried submitting a large file 3 times in the [qasp environment](https://tdp-frontend-qasp.app.cloud.gov/data-files), but I was unable to get a successful submission + completed parsing result. Let me know if I should try this a different way. 




---
```
#try 1: Successfully submitted >  worker exited early > transmitted to ACF titan server
13:29:06.173: [APP/PROC/WEB.0] [2023-07-11 17:29:06,172 INFO parser_task.py::parse:L19 :  DataFile parsing started for file ADS.E2J.NDM1.TS26
13:34:26.768: [APP/PROC/WEB.0] [2023-07-11 17:34:26,767: ERROR/MainProcess] Process 'ForkPoolWorker-1' pid:564 exited with 'signal 9 (SIGKILL)'
13:34:26.783: [APP/PROC/WEB.0] [2023-07-11 17:34:26,782: ERROR/MainProcess] Task handler raised error: WorkerLostError('Worker exited prematurely: signal 9 (SIGKILL) Job: 0.')
13:34:26.783: [APP/PROC/WEB.0] Traceback (most recent call last):
13:34:26.783: [APP/PROC/WEB.0]   File "/home/vcap/deps/1/python/lib/python3.10/site-packages/billiard/pool.py", line 1265, in mark_as_worker_lost
13:34:26.783: [APP/PROC/WEB.0]     raise WorkerLostError(
13:34:26.783: [APP/PROC/WEB.0] billiard.exceptions.WorkerLostError: Worker exited prematurely: signal 9 (SIGKILL) Job: 0.
13:34:34.802: [APP/PROC/WEB.0] [2023-07-11 17:34:34,801 INFO sftp_task.py::upload:L97 :  File ADS.E2J.NDM1.TS26 has been successfully uploaded to acftitan.cit.nih.gov
13:34:34.802: [APP/PROC/WEB.0] [2023-07-11 17:34:34,801: INFO/ForkPoolWorker-2] File ADS.E2J.NDM1.TS26 has been successfully uploaded to acftitan.cit.nih.gov
```
![2486p1](https://github.com/raft-tech/TANF-app/assets/63075587/15e92aab-30b3-4fde-9ee4-1cc863b94d43)
----
```
#  try #2: submission unsuccessful: 400 response returned on frontend
13:36:09.438: [APP/PROC/WEB.0] 10.255.218.207 - - [11/Jul/2023:17:36:09 +0000] "GET /v1/data_files/?year=2023&quarter=Q2&stt=23&file_type=tanf HTTP/1.0" 200 441 "https://tdp-frontend-qasp.app.cloud.gov/data-files" "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/114.0.0.0 Safari/537.36"
13:36:09.558: [APP/PROC/WEB.0] 10.255.218.207 - - [11/Jul/2023:17:36:09 +0000] "GET /v1/data_files/?year=2023&quarter=Q2&stt=23&file_type=tanf HTTP/1.0" 200 441 "https://tdp-frontend-qasp.app.cloud.gov/data-files" "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/114.0.0.0 Safari/537.36"
13:36:12.308: [APP/PROC/WEB.0] [2023-07-11 17:36:12 +0000] [567] [DEBUG] Closing connection. 
13:36:42.337: [APP/PROC/WEB.0] [2023-07-11 17:36:42 +0000] [569] [DEBUG] Closing connection. 
13:36:47.632: [APP/PROC/WEB.0] [2023-07-11 17:36:47 +0000] [569] [DEBUG] POST /v1/data_files/
13:36:47.977: [APP/PROC/WEB.0] WARNING Bad Request: /v1/data_files/
13:36:47.977: [APP/PROC/WEB.0] 10.255.218.207 - - [11/Jul/2023:17:36:47 +0000] "POST /v1/data_files/ HTTP/1.0" 400 128 "https://tdp-frontend-qasp.app.cloud.gov/data-files" "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/114.0.0.0 Safari/537.36"
```
----
```
#try 3: Successfully submitted >  worker exited early > transmitted to ACF titan server
0.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/114.0.0.0 Safari/537.36"
13:38:36.331: [APP/PROC/WEB.0] [2023-07-11 17:38:36,330 INFO views.py::create:L80 :  Submitted upload task to redis for datafile 64.
13:38:36.331: [APP/PROC/WEB.0] Submitted upload task to redis for datafile 64.
13:38:36.344: [APP/PROC/WEB.0] [2023-07-11 17:38:36,343 INFO parser_task.py::parse:L19 :  DataFile parsing started for file ADS.E2J.NDM1.TS26
13:38:36.344: [APP/PROC/WEB.0] [2023-07-11 17:38:36,343: INFO/ForkPoolWorker-2] DataFile parsing started for file ADS.E2J.NDM1.TS26
13:38:36.607: [APP/PROC/WEB.0] [2023-07-11 17:38:36 +0000] [567] [DEBUG] POST /v1/logs/
13:38:36.633: [APP/PROC/WEB.0] 10.255.218.207 - - [11/Jul/2023:17:38:36 +0000] "POST /v1/logs/ HTTP/1.0" 200 9 "https://tdp-frontend-qasp.app.cloud.gov/data-files" "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/114.0.0.0 Safari/537.36"
13:38:42.450: [APP/PROC/WEB.0] [2023-07-11 17:38:42 +0000] [570] [DEBUG] Closing connection. 
13:39:12.478: [APP/PROC/WEB.0] [2023-07-11 17:39:12 +0000] [569] [DEBUG] Closing connection. 
13:39:42.506: [APP/PROC/WEB.0] [2023-07-11 17:39:42 +0000] [571] [DEBUG] Closing connection. 
13:40:12.535: [APP/PROC/WEB.0] [2023-07-11 17:40:12 +0000] [571] [DEBUG] Closing connection. 
13:40:42.563: [APP/PROC/WEB.0] [2023-07-11 17:40:42 +0000] [571] [DEBUG] Closing connection. 
13:41:12.590: [APP/PROC/WEB.0] [2023-07-11 17:41:12 +0000] [571] [DEBUG] Closing connection. 
13:41:17.382: [APP/PROC/WEB.0] [2023-07-11 17:41:17,381: ERROR/MainProcess] Process 'ForkPoolWorker-2' pid:614 exited with 'signal 9 (SIGKILL)'
13:41:17.393: [APP/PROC/WEB.0] [2023-07-11 17:41:17,393: ERROR/MainProcess] Task handler raised error: WorkerLostError('Worker exited prematurely: signal 9 (SIGKILL) Job: 3.')
13:41:17.393: [APP/PROC/WEB.0] Traceback (most recent call last):
13:41:17.393: [APP/PROC/WEB.0]   File "/home/vcap/deps/1/python/lib/python3.10/site-packages/billiard/pool.py", line 1265, in mark_as_worker_lost
13:41:17.393: [APP/PROC/WEB.0]     raise WorkerLostError(
13:41:17.393: [APP/PROC/WEB.0] billiard.exceptions.WorkerLostError: Worker exited prematurely: signal 9 (SIGKILL) Job: 3.
13:41:25.110: [APP/PROC/WEB.0] [2023-07-11 17:41:25,109 INFO sftp_task.py::upload:L97 :  File ADS.E2J.NDM1.TS26 has been successfully uploaded to acftitan.cit.nih.gov
13:41:25.110: [APP/PROC/WEB.0] [2023-07-11 17:41:25,109: INFO/ForkPoolWorker-3] File ADS.E2J.NDM1.TS26 has been successfully uploaded to acftitan.cit.nih.gov
```
![2486p2](https://github.com/raft-tech/TANF-app/assets/63075587/a519f6f9-be2d-4f02-acce-3ff0c2acd040)

elipe17 commented:
> @elipe17 testing notes below. I tried submitting a large file 3 times in the [qasp environment](https://tdp-frontend-qasp.app.cloud.gov/data-files), but I was unable to get a successful submission + completed parsing result. Let me know if I should try this a different way.
> 
> ```
> #try 1: Successfully submitted >  worker exited early > transmitted to ACF titan server
> 13:29:06.173: [APP/PROC/WEB.0] [2023-07-11 17:29:06,172 INFO parser_task.py::parse:L19 :  DataFile parsing started for file ADS.E2J.NDM1.TS26
> 13:34:26.768: [APP/PROC/WEB.0] [2023-07-11 17:34:26,767: ERROR/MainProcess] Process 'ForkPoolWorker-1' pid:564 exited with 'signal 9 (SIGKILL)'
> 13:34:26.783: [APP/PROC/WEB.0] [2023-07-11 17:34:26,782: ERROR/MainProcess] Task handler raised error: WorkerLostError('Worker exited prematurely: signal 9 (SIGKILL) Job: 0.')
> 13:34:26.783: [APP/PROC/WEB.0] Traceback (most recent call last):
> 13:34:26.783: [APP/PROC/WEB.0]   File "/home/vcap/deps/1/python/lib/python3.10/site-packages/billiard/pool.py", line 1265, in mark_as_worker_lost
> 13:34:26.783: [APP/PROC/WEB.0]     raise WorkerLostError(
> 13:34:26.783: [APP/PROC/WEB.0] billiard.exceptions.WorkerLostError: Worker exited prematurely: signal 9 (SIGKILL) Job: 0.
> 13:34:34.802: [APP/PROC/WEB.0] [2023-07-11 17:34:34,801 INFO sftp_task.py::upload:L97 :  File ADS.E2J.NDM1.TS26 has been successfully uploaded to acftitan.cit.nih.gov
> 13:34:34.802: [APP/PROC/WEB.0] [2023-07-11 17:34:34,801: INFO/ForkPoolWorker-2] File ADS.E2J.NDM1.TS26 has been successfully uploaded to acftitan.cit.nih.gov
> ```
> 
> ## ![2486p1](https://user-images.githubusercontent.com/63075587/252751031-15e92aab-30b3-4fde-9ee4-1cc863b94d43.PNG)
> ```
> #  try #2: submission unsuccessful: 400 response returned on frontend
> 13:36:09.438: [APP/PROC/WEB.0] 10.255.218.207 - - [11/Jul/2023:17:36:09 +0000] "GET /v1/data_files/?year=2023&quarter=Q2&stt=23&file_type=tanf HTTP/1.0" 200 441 "https://tdp-frontend-qasp.app.cloud.gov/data-files" "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/114.0.0.0 Safari/537.36"
> 13:36:09.558: [APP/PROC/WEB.0] 10.255.218.207 - - [11/Jul/2023:17:36:09 +0000] "GET /v1/data_files/?year=2023&quarter=Q2&stt=23&file_type=tanf HTTP/1.0" 200 441 "https://tdp-frontend-qasp.app.cloud.gov/data-files" "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/114.0.0.0 Safari/537.36"
> 13:36:12.308: [APP/PROC/WEB.0] [2023-07-11 17:36:12 +0000] [567] [DEBUG] Closing connection. 
> 13:36:42.337: [APP/PROC/WEB.0] [2023-07-11 17:36:42 +0000] [569] [DEBUG] Closing connection. 
> 13:36:47.632: [APP/PROC/WEB.0] [2023-07-11 17:36:47 +0000] [569] [DEBUG] POST /v1/data_files/
> 13:36:47.977: [APP/PROC/WEB.0] WARNING Bad Request: /v1/data_files/
> 13:36:47.977: [APP/PROC/WEB.0] 10.255.218.207 - - [11/Jul/2023:17:36:47 +0000] "POST /v1/data_files/ HTTP/1.0" 400 128 "https://tdp-frontend-qasp.app.cloud.gov/data-files" "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/114.0.0.0 Safari/537.36"
> ```
> 
> ```
> #try 3: Successfully submitted >  worker exited early > transmitted to ACF titan server
> 0.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/114.0.0.0 Safari/537.36"
> 13:38:36.331: [APP/PROC/WEB.0] [2023-07-11 17:38:36,330 INFO views.py::create:L80 :  Submitted upload task to redis for datafile 64.
> 13:38:36.331: [APP/PROC/WEB.0] Submitted upload task to redis for datafile 64.
> 13:38:36.344: [APP/PROC/WEB.0] [2023-07-11 17:38:36,343 INFO parser_task.py::parse:L19 :  DataFile parsing started for file ADS.E2J.NDM1.TS26
> 13:38:36.344: [APP/PROC/WEB.0] [2023-07-11 17:38:36,343: INFO/ForkPoolWorker-2] DataFile parsing started for file ADS.E2J.NDM1.TS26
> 13:38:36.607: [APP/PROC/WEB.0] [2023-07-11 17:38:36 +0000] [567] [DEBUG] POST /v1/logs/
> 13:38:36.633: [APP/PROC/WEB.0] 10.255.218.207 - - [11/Jul/2023:17:38:36 +0000] "POST /v1/logs/ HTTP/1.0" 200 9 "https://tdp-frontend-qasp.app.cloud.gov/data-files" "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/114.0.0.0 Safari/537.36"
> 13:38:42.450: [APP/PROC/WEB.0] [2023-07-11 17:38:42 +0000] [570] [DEBUG] Closing connection. 
> 13:39:12.478: [APP/PROC/WEB.0] [2023-07-11 17:39:12 +0000] [569] [DEBUG] Closing connection. 
> 13:39:42.506: [APP/PROC/WEB.0] [2023-07-11 17:39:42 +0000] [571] [DEBUG] Closing connection. 
> 13:40:12.535: [APP/PROC/WEB.0] [2023-07-11 17:40:12 +0000] [571] [DEBUG] Closing connection. 
> 13:40:42.563: [APP/PROC/WEB.0] [2023-07-11 17:40:42 +0000] [571] [DEBUG] Closing connection. 
> 13:41:12.590: [APP/PROC/WEB.0] [2023-07-11 17:41:12 +0000] [571] [DEBUG] Closing connection. 
> 13:41:17.382: [APP/PROC/WEB.0] [2023-07-11 17:41:17,381: ERROR/MainProcess] Process 'ForkPoolWorker-2' pid:614 exited with 'signal 9 (SIGKILL)'
> 13:41:17.393: [APP/PROC/WEB.0] [2023-07-11 17:41:17,393: ERROR/MainProcess] Task handler raised error: WorkerLostError('Worker exited prematurely: signal 9 (SIGKILL) Job: 3.')
> 13:41:17.393: [APP/PROC/WEB.0] Traceback (most recent call last):
> 13:41:17.393: [APP/PROC/WEB.0]   File "/home/vcap/deps/1/python/lib/python3.10/site-packages/billiard/pool.py", line 1265, in mark_as_worker_lost
> 13:41:17.393: [APP/PROC/WEB.0]     raise WorkerLostError(
> 13:41:17.393: [APP/PROC/WEB.0] billiard.exceptions.WorkerLostError: Worker exited prematurely: signal 9 (SIGKILL) Job: 3.
> 13:41:25.110: [APP/PROC/WEB.0] [2023-07-11 17:41:25,109 INFO sftp_task.py::upload:L97 :  File ADS.E2J.NDM1.TS26 has been successfully uploaded to acftitan.cit.nih.gov
> 13:41:25.110: [APP/PROC/WEB.0] [2023-07-11 17:41:25,109: INFO/ForkPoolWorker-3] File ADS.E2J.NDM1.TS26 has been successfully uploaded to acftitan.cit.nih.gov
> ```
> 
> ![2486p2](https://user-images.githubusercontent.com/63075587/252751039-a519f6f9-be2d-4f02-acce-3ff0c2acd040.PNG)

@ADPennington what was the size of the file? We have unit tests that execute on a 50MB file with no issue. Theoretically the size of the file shouldnt matter, but I want to have it as a data point for debugging.

ADPennington commented:
@elipe17 50,439KB

elipe17 commented:
> WorkerLostError

@ADPennington  I spent most of yesterday doing another memory analysis on this branch and `develop`. I have come to the conclusion that the issue is primarily an infrastructure issue. I ran the 50MB file on this branch and `develop` and compared the memory and cpu consumption. On average (over five runs), `develop` consumed 100MB more memory than this branch while parsing the 50MB file. With that said, I dont think this PR is introducing a regression but is even more so optimal than `develop` based on my empirical analysis. I think what is happening is there are occasional short lived memory spikes in the deployed environment which causes the celery worker to be killed by the Linux OOM Killer. This could potentially be attributed to a memory leak somewhere in either the app code or our dependencies code. For the time being I don't think there is an immediate remedy for this issue other than capping max file size or increasing the memory capacity in the deployed environments. Please let me know how you would like to proceed and I would love to answer any questions you might have!

elipe17 committed:
- Updating based on comments.
- Added base class to keep things a little more DRY

elipe17 committed:
Merge branch '2579-latest-data' into 2486-parser-performance

elipe17 committed:
- Refactored parsing to avoid custom transactions by leveraging the FK on the records. Rollbacks are a lot slower now, but hopefully that will happen much less versus successful parses.

elipe17 committed:
- small optimization for the rollback

elipe17 committed:
- Fix lint errors

elipe17 committed:
- Removing commented code

elipe17 committed:
Merge branch 'develop' of https://github.com/raft-tech/TANF-app into 2579-latest-data

elipe17 committed:
- Installing build essentials

elipe17 committed:
Revert "- Installing build essentials"

This reverts commit 782a5f7d687e60965a9100520f027b9678dbac35.

elipe17 committed:
- adding setup tools and wheel

elipe17 committed:
Revert "- adding setup tools and wheel"

This reverts commit f529728811fba242132b7c42f9e9e09d6037fa70.

elipe17 committed:
- Updating dependencies to get around pep issue

elipe17 committed:
- Pin factoryboy
- fix lint error

elipe17 committed:
Merge branch '2579-latest-data' into 2486-parser-performance

elipe17 commented:
The latest changes I made here do not resolve the issue where celery workers are crashing when parsing a large (50MB) file. This is not a regression on this branch and falls outside of the scope of this issue. Also, 2579 needs to merge prior to this being merged since it contains 2579.

elipe17 committed:
Merge branch 'develop' of https://github.com/raft-tech/TANF-app into 2486-parser-performance

elipe17 committed:
- Updating manifest

elipe17 commented:
@ADPennington After upping the memory to 2GB we successfully parsed the 50MB file. You should be able to re-deploy to qasp and duplicate the test. Remember to add a `.txt` extension to the file if it doesnt have one so you dont get stuck like me :laughing:

elipe17 commented:
After upping the memory in the deployed environments to 2GB we can successfully parse the 50MB file without an OOM error. Once the Raft review passes it will be ready for you @ADPennington.

elipe17 commented:
@andrew-jameson or @raftmsohani could one of you review this again so I can get it into qasp?

raftmsohani reviewed (approved):
LGTM!

elipe17 commented:
@ADPennington This is ready for your review again. I don't think it is actually deployed in qasp even though the label is there.

ADPennington reviewed (approved):
testing with large file looks good. also great to see some of the errors from the file. thanks @elipe17 🚀
